core/src/main/scala/kafka/server/AlterIsrChannelManager.scala

```diff
@@ -0,0 +1,132 @@
+  def shutdown(): Unit
+}
+
+case class AlterIsrItem(topicPartition: TopicPartition, leaderAndIsr: LeaderAndIsr)
+
+class AlterIsrChannelManagerImpl(val controllerChannelManager: BrokerToControllerChannelManager,
+                                 val zkClient: KafkaZkClient,
+                                 val scheduler: Scheduler,
+                                 val brokerId: Int,
+                                 val brokerEpoch: Long) extends AlterIsrChannelManager with Logging with KafkaMetricsGroup {
+
+  private val pendingIsrUpdates: mutable.Map[TopicPartition, AlterIsrItem] = new mutable.HashMap[TopicPartition, AlterIsrItem]()
+  private val lastIsrChangeMs = new AtomicLong(0)
+  private val lastIsrPropagationMs = new AtomicLong(0)
+
+  @volatile private var scheduledRequest: Option[ScheduledFuture[_]] = None
+
+  override def enqueueIsrUpdate(alterIsrItem: AlterIsrItem): Unit = {
+    pendingIsrUpdates synchronized {
+      pendingIsrUpdates(alterIsrItem.topicPartition) = alterIsrItem
+      lastIsrChangeMs.set(System.currentTimeMillis())
+      // Rather than sending right away, we'll delay at most 50ms to allow for batching of ISR changes happening
+      // in fast succession
+      if (scheduledRequest.isEmpty) {
+        scheduledRequest = Some(scheduler.schedule("propagate-alter-isr", propagateIsrChanges, 50, -1, TimeUnit.MILLISECONDS))
+      }
+    }
+  }
+
+  override def clearPending(topicPartition: TopicPartition): Unit = {
+    pendingIsrUpdates synchronized {
+      // when we get a new LeaderAndIsr, we clear out any pending requests
+      pendingIsrUpdates.remove(topicPartition)
+    }
+  }
+
+  override def startup(): Unit = {
+    controllerChannelManager.start()
+  }
+
+  override def shutdown(): Unit = {
+    controllerChannelManager.shutdown()
+  }
+
+  private def propagateIsrChanges(): Unit = {
+    val now = System.currentTimeMillis()
+    pendingIsrUpdates synchronized {
+      if (pendingIsrUpdates.nonEmpty) {
+        // Max ISRs to send?
+        val message = new AlterIsrRequestData()
+          .setBrokerId(brokerId)
+          .setBrokerEpoch(brokerEpoch)
+          .setTopics(new util.ArrayList())
+
+        pendingIsrUpdates.values.groupBy(_.topicPartition.topic()).foreachEntry((topic, items) => {
+          val topicPart = new AlterIsrRequestTopics()
+            .setName(topic)
+            .setPartitions(new util.ArrayList())
+          message.topics().add(topicPart)
+          items.foreach(item => {
+            topicPart.partitions().add(new AlterIsrRequestPartitions()
+              .setPartitionIndex(item.topicPartition.partition())
+              .setLeaderId(item.leaderAndIsr.leader)
+              .setLeaderEpoch(item.leaderAndIsr.leaderEpoch)
+              .setNewIsr(item.leaderAndIsr.isr.map(Integer.valueOf).asJava)
+              .setCurrentIsrVersion(item.leaderAndIsr.zkVersion)
+            )
+          })
+        })
+
+        def responseHandler(response: ClientResponse): Unit = {
```

Comment:
I think the basic approach here is to ignore successful responses and wait for the `LeaderAndIsr` update. I am wondering how we should handle the case when the update failed. Say for example that our update fails with the INVALID_VERSION error. Inside `Partition`, we will still have the pendingIsr set. Do we need to clear it? How about other errors?

Reply:
Turns out we do need to clear it. In the common case, the controller will still send out LeaderAndIsr following a failed ISR update. However, during testing I did find a case where a LeaderAndIsr failed to go out leading to a stuck leader.

* ISR = {1,2}
* Broker 3 comes up
* Leader sends ISR = {1,2,3}
* Broker 3 goes down
* Controller handles broker shutdown
* Controller handles AlterIsr, resulting LeaderAndIsr batch fails to send because broker 3 is offline
* Broker 3 comes up, leader still waiting on LeaderAndIsr and never get is -- ISR gets stuck

Reply:
I found a race during the system tests when a broker is shutting down. The controller handles the shut down before it handles an AlterIsr. If the proposed ISR includes the now-offline replica, the controller refuses to update that ISR change and returns an error for that partition. It then sends out the current LeaderAndIsr.

The problem is that the broker ignores this LeaderAndIsr since it has the same leader epoch. This is easy enough to fix, we can bump the leader epoch in the controller (and ZK) before sending it out.

However, there's still the case of failing to update ZK. I think we should probably treat this the same way as an offline replica. If we simply return an error in AlterIsr response and let the leader reset the pending ISR state, the leader will just retry with stale metadata and the update will fail again. 

I think in all these error cases we must bump the leader epoch to force the leader to accept the new LeaderAndIsr. Thoughts?